### PR TITLE
[BEAM-115] Fn API support for Python

### PR DESCRIPTION
**Pull request by robertwb**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: sdks/python/apache_beam/runners/portability/fn_api_runner_test.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import logging
+import unittest
+
+import apache_beam as beam
+from apache_beam.runners.portability import fn_api_runner
+from apache_beam.runners.portability import maptask_executor_runner
+
+
+class FnApiRunnerTest(maptask_executor_runner.MapTaskExecutorRunner):
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_test.py]: did you mean `worker_runner_base_test.WorkerRunnerBaseTest` as the base class?

File: sdks/python/apache_beam/runners/worker/log_handler.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  # Maximum number of log entries in a single stream request.
+  _MAX_BATCH_SIZE = 1000
+  # Used to indicate the end of stream.
+  _FINISHED = object()
+
+  # Mapping from logging levels to LogEntry levels.
+  LOG_LEVEL_MAP = {
+      logging.FATAL: beam_fn_api_pb2.LogEntry.CRITICAL,
+      logging.ERROR: beam_fn_api_pb2.LogEntry.ERROR,
+      logging.WARNING: beam_fn_api_pb2.LogEntry.ERROR,
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/log_handler.py]: `beam_fn_api_pb2.LogEntry.WARN`?

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
+"""A PipelineRunner using the SDK harness.
+"""
+import collections
+import json
+import logging
+import Queue as queue
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: For consistency. The name of the class is `Queue`. (This wart was fixed in Python 3.)

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+            and isinstance(operation.source.source.default_output_coder(),
+                           WindowedValueCoder)):
+          output_stream = create_OutputStream()
+          element_coder = (
+              operation.source.source.default_output_coder().get_impl())
+          # Re-encode the elements in the nested context and
+          # concatenate them together
+          for element in operation.source.source.read(None):
+            element_coder.encode_to_stream(element, output_stream, True)
+          target_name = '%s' % self._next_uid()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: Fixed.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_test.py (hunk 1 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_test.py]: This was renamed. Fixed the reference below.

File: sdks/python/apache_beam/runners/worker/log_handler.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/log_handler.py]: Done.

**Comment by robertwb** [comment]
R: @aaltay 

I think it's all in shape now, including licenses (we crossed on that one) but it's a fair chunk of code, so awaiting your comments. I'll stop force-pushing now to avoid loss of context.

File: sdks/python/apache_beam/utils/portpicker.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import contextlib
+import socket
+
+
+def pick_unused_port():
```
Comment (robertwb) [inline sdks/python/apache_beam/utils/portpicker.py]: Done.

File: sdks/python/apache_beam/runners/worker/statesampler_test.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      time.sleep(0.1)
+      with sampler.scoped_state('stateb'):
+        time.sleep(0.2 / 2)
+        with sampler.scoped_state('statec'):
+          time.sleep(0.3)
+        time.sleep(0.2 / 2)
+    sampler.stop()
+    sampler.commit_counters()
+
+    # Test that sampled state timings are within 5% of expected values.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_test.py]: Yeah. It flakes about 2% of the time with 5% margin of error (and did so even without using locks). Full seconds would make it stabler, but make the test more expensive. Another option would be to require within 25%, and if not within 5% retry. 

Future PR (though I fixed the comment).

File: sdks/python/apache_beam/runners/worker/statesampler_test.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    state_transition_count = sampler.get_info().transition_count
+    overhead_us = 1000000.0 * elapsed_time / state_transition_count
+    logging.info('Overhead per transition: %fus', overhead_us)
+    # Conservative upper bound on overhead in microseconds (we expect this to
+    # take 0.17us when compiled in opt mode or 0.48 us when compiled with in
+    # debug mode).
+    self.assertLess(overhead_us, 10.0)
+
+
+try:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_test.py]: Done.

File: sdks/python/apache_beam/runners/worker/statesampler_fake.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+class StateSampler(object):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_fake.py]: For tests that need to pass a StateSampler to construct objects in pure Python mode.

File: sdks/python/apache_beam/runners/worker/statesampler.pyx [hunk 9]
```diff
@@ -0,0 +1,234 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.pyx]: Fixed all the .pyx and .pxd ones.

File: sdks/python/apache_beam/runners/worker/sdk_worker_test.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+          logging.info("All %s instructions finished.", len(self.requests))
+          return
+    raise RuntimeError("Missing responses: %s" %
+                       (self.instruction_ids - set(self.responses.keys())))
+
+
+class SdkWorkerTest(unittest.TestCase):
+
+  def test_fn_registration(self):
+
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/sdk_worker_test.py]: Done.

File: sdks/python/setup.py [hunk 11]
```diff
@@ -125,8 +125,11 @@ def get_version():
     ext_modules=cythonize([
         '**/*.pyx',
```
Comment (aaltay) [inline sdks/python/setup.py]: Would this include `runners/worker/statesampler.pyx` ?

**Comment by aaltay** [comment]
@robertwb it LGTM, I do not have additional comments except for the open ones.

I have read it all once, but given the size I did not go through the details of every bit. Do you want me to spend more time on any specific part?

File: sdks/python/setup.py (hunk 11 above)
Comment (robertwb) [inline sdks/python/setup.py]: Yes.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+    def __init__(self):
+      self.state_handler = FnApiRunner.SimpleState()
+      self.control_server = grpc.server(
+          futures.ThreadPoolExecutor(max_workers=10))
+      self.control_port = portpicker.pick_unused_port()
+      self.control_server.add_insecure_port('[::]:%s' % self.control_port)
+
+      self.data_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+      self.data_port = portpicker.pick_unused_port()
+      self.data_server.add_insecure_port('[::]:%s' % self.data_port)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: Oh, that would be much nicer. I'll see if I can get this to work.

**Comment by robertwb** [comment]
Given that most of this is (already reviewed and tested) worker code, I think that's sufficient (and still quite the job, thanks!)

**Comment by robertwb** [comment]
Jenkins: retest this please

**Pull request closed without merging**